File: config/migrate.js
import { query } from './db.js';

const tables = {
  users: {
    columns: [
      { name: 'id', def: 'INT AUTO_INCREMENT PRIMARY KEY' },
      { name: 'email', def: 'VARCHAR(255) NOT NULL' },
      { name: 'username', def: 'VARCHAR(255) NOT NULL' },
      { name: 'password_hash', def: 'VARCHAR(255) NOT NULL' },
      { name: 'ptero_user_id', def: 'INT' },
      { name: 'ptero_uuid', def: 'VARCHAR(255)' },
      { name: 'first_name', def: 'VARCHAR(255)' },
      { name: 'last_name', def: 'VARCHAR(255)' },
      { name: 'password_set', def: 'TINYINT(1) NOT NULL DEFAULT 0' },
      { name: 'is_admin', def: 'TINYINT(1) NOT NULL DEFAULT 0' },
      { name: 'restricted', def: 'TINYINT(1) NOT NULL DEFAULT 0' },
      { name: 'auth_restricted', def: 'TINYINT(1) NOT NULL DEFAULT 0' },
      { name: 'token_version', def: 'INT NOT NULL DEFAULT 0' },
      { name: 'avatar', def: 'VARCHAR(255) DEFAULT NULL' },
      { name: 'ptero_client_api_key', def: 'VARCHAR(255) DEFAULT NULL' },
      { name: 'email_verified', def: 'TINYINT(1) NOT NULL DEFAULT 0' },
      { name: 'verification_token', def: 'VARCHAR(64) DEFAULT NULL' },
      { name: 'verification_token_expires', def: 'DATETIME DEFAULT NULL' },
      { name: 'created_at', def: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP' },
    ],
  },
  server_meta: {
    columns: [
      { name: 'id', def: 'INT AUTO_INCREMENT PRIMARY KEY' },
      { name: 'ptero_server_id', def: 'INT NOT NULL' },
      { name: 'user_id', def: 'INT NOT NULL' },
      { name: 'created_at', def: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP' },
      { name: 'expires_at', def: 'TIMESTAMP NOT NULL' },
      { name: 'status', def: "ENUM('active', 'suspended', 'expired') DEFAULT 'active'" },
      { name: 'suspend_reason', def: 'TEXT DEFAULT NULL' },
      { name: 'suspended_by', def: "VARCHAR(20) DEFAULT NULL" },
    ],
  },
  user_ips: {
    columns: [
      { name: 'id', def: 'INT AUTO_INCREMENT PRIMARY KEY' },
      { name: 'user_id', def: 'INT NOT NULL' },
      { name: 'ip_address', def: 'VARCHAR(45) NOT NULL' },
      { name: 'created_at', def: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP' },
    ],
  },
  activity_log: {
    columns: [
      { name: 'id', def: 'INT AUTO_INCREMENT PRIMARY KEY' },
      { name: 'user_id', def: 'INT NOT NULL' },
      { name: 'action', def: 'VARCHAR(50) NOT NULL' },
      { name: 'details', def: 'VARCHAR(255) DEFAULT \'\'' },
      { name: 'server_id', def: 'INT DEFAULT NULL' },
      { name: 'created_at', def: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP' },
    ],
  },
  nests: {
    columns: [
      { name: 'id', def: 'INT AUTO_INCREMENT PRIMARY KEY' },
      { name: 'ptero_nest_id', def: 'INT NOT NULL UNIQUE' },
      { name: 'name', def: 'VARCHAR(255) NOT NULL' },
      { name: 'logo', def: 'VARCHAR(255) DEFAULT NULL' },
      { name: 'description', def: 'TEXT DEFAULT NULL' },
      { name: 'unavailable', def: 'TINYINT(1) NOT NULL DEFAULT 0' },
      { name: 'created_at', def: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP' },
    ],
  },
  egg_resources: {
    columns: [
      { name: 'id', def: 'INT AUTO_INCREMENT PRIMARY KEY' },
      { name: 'ptero_nest_id', def: 'INT NOT NULL' },
      { name: 'ptero_egg_id', def: 'INT NOT NULL' },
      { name: 'logo', def: 'VARCHAR(255) DEFAULT NULL' },
      { name: 'cpu_limit', def: 'INT DEFAULT NULL' },
      { name: 'memory_limit', def: 'INT DEFAULT NULL' },
      { name: 'disk_limit', def: 'INT DEFAULT NULL' },
      { name: 'unavailable', def: 'TINYINT(1) NOT NULL DEFAULT 0' },
      { name: 'created_at', def: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP' },
      { name: 'updated_at', def: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP' },
    ],
  },
  notifications: {
    columns: [
      { name: 'id', def: 'INT AUTO_INCREMENT PRIMARY KEY' },
      { name: 'user_id', def: 'INT NOT NULL' },
      { name: 'title', def: 'VARCHAR(255) NOT NULL' },
      { name: 'message', def: 'TEXT NOT NULL' },
      { name: 'type', def: "VARCHAR(20) NOT NULL DEFAULT 'info'" },
      { name: 'link', def: 'VARCHAR(255) DEFAULT NULL' },
      { name: 'is_read', def: 'TINYINT(1) NOT NULL DEFAULT 0' },
      { name: 'created_at', def: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP' },
    ],
  },
  passkeys: {
    columns: [
      { name: 'id', def: 'INT AUTO_INCREMENT PRIMARY KEY' },
      { name: 'user_id', def: 'INT NOT NULL' },
      { name: 'credential_id', def: 'VARCHAR(512) NOT NULL' },
      { name: 'public_key', def: 'TEXT NOT NULL' },
      { name: 'counter', def: 'INT NOT NULL DEFAULT 0' },
      { name: 'transports', def: 'VARCHAR(255) DEFAULT NULL' },
      { name: 'name', def: 'VARCHAR(255) DEFAULT NULL' },
      { name: 'created_at', def: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP' },
    ],
  },
  node_settings: {
    columns: [
      { name: 'id', def: 'INT AUTO_INCREMENT PRIMARY KEY' },
      { name: 'ptero_node_id', def: 'INT NOT NULL UNIQUE' },
      { name: 'unavailable', def: 'TINYINT(1) NOT NULL DEFAULT 0' },
      { name: 'created_at', def: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP' },
      { name: 'updated_at', def: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP' },
    ],
  },

};

function escapeId(id) {
  return '`' + String(id).replace(/`/g, '``') + '`';
}

function escapeLike(id) {
  return String(id).replace(/[%_\\]/g, '\\$&');
}

const constraints = [
  { table: 'server_meta', sql: 'ALTER TABLE server_meta ADD INDEX idx_expires (expires_at)', name: 'idx_expires' },
  { table: 'server_meta', sql: 'ALTER TABLE server_meta ADD INDEX idx_user (user_id)', name: 'idx_user' },
  { table: 'server_meta', sql: 'ALTER TABLE server_meta ADD INDEX idx_status (status)', name: 'idx_status' },
  { table: 'user_ips', sql: 'ALTER TABLE user_ips ADD INDEX idx_ip (ip_address)', name: 'idx_ip' },
  { table: 'user_ips', sql: 'ALTER TABLE user_ips ADD INDEX idx_user (user_id)', name: 'idx_user' },
  { table: 'user_ips', sql: 'ALTER TABLE user_ips ADD CONSTRAINT fk_user_ips_user FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE', name: 'fk_user_ips_user' },
  { table: 'activity_log', sql: 'ALTER TABLE activity_log ADD INDEX idx_activity_user (user_id)', name: 'idx_activity_user' },
  { table: 'activity_log', sql: 'ALTER TABLE activity_log ADD INDEX idx_activity_created (created_at)', name: 'idx_activity_created' },
  { table: 'egg_resources', sql: 'ALTER TABLE egg_resources ADD UNIQUE INDEX idx_egg_resources_nest_egg (ptero_nest_id, ptero_egg_id)', name: 'idx_egg_resources_nest_egg' },
  { table: 'notifications', sql: 'ALTER TABLE notifications ADD INDEX idx_notif_user (user_id)', name: 'idx_notif_user' },
  { table: 'notifications', sql: 'ALTER TABLE notifications ADD INDEX idx_notif_user_read (user_id, is_read)', name: 'idx_notif_user_read' },
  { table: 'notifications', sql: 'ALTER TABLE notifications ADD INDEX idx_notif_created (created_at)', name: 'idx_notif_created' },
  { table: 'notifications', sql: 'ALTER TABLE notifications ADD CONSTRAINT fk_notif_user FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE', name: 'fk_notif_user' },
  { table: 'users', sql: 'ALTER TABLE users ADD INDEX idx_email (email)', name: 'idx_user_email' },
  { table: 'users', sql: 'ALTER TABLE users ADD INDEX idx_username (username)', name: 'idx_user_username' },
  { table: 'activity_log', sql: 'ALTER TABLE activity_log ADD INDEX idx_action (action)', name: 'idx_activity_action' },
  { table: 'activity_log', sql: 'ALTER TABLE activity_log ADD CONSTRAINT fk_activity_user FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE', name: 'fk_activity_user' },
  { table: 'server_meta', sql: 'ALTER TABLE server_meta ADD INDEX idx_ptero_server (ptero_server_id)', name: 'idx_ptero_server' },
  { table: 'server_meta', sql: 'ALTER TABLE server_meta ADD CONSTRAINT fk_server_meta_user FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE', name: 'fk_server_meta_user' },
  { table: 'passkeys', sql: 'ALTER TABLE passkeys ADD INDEX idx_passkey_user (user_id)', name: 'idx_passkey_user' },
  { table: 'passkeys', sql: 'ALTER TABLE passkeys ADD INDEX idx_passkey_credential (credential_id(255))', name: 'idx_passkey_credential' },
  { table: 'passkeys', sql: 'ALTER TABLE passkeys ADD CONSTRAINT fk_passkey_user FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE', name: 'fk_passkey_user' },
  { table: 'node_settings', sql: 'ALTER TABLE node_settings ADD INDEX idx_node_settings_node (ptero_node_id)', name: 'idx_node_settings_node' },
];

export async function migrate() {
  for (const [table, schema] of Object.entries(tables)) {
    try {
      const safeTable = escapeId(table);
      const exists = await query(`SHOW TABLES LIKE ?`, [escapeLike(table)]);
      if (exists.length === 0) {
        const colDefs = schema.columns.map(c => `${escapeId(c.name)} ${c.def}`).join(', ');
        await query(`CREATE TABLE ${safeTable} (${colDefs})`);
        console.log(`Created table: ${table}`);
        continue;
      }

      for (const col of schema.columns) {
        try {
          const safeCol = escapeId(col.name);
          const cols = await query(`SHOW COLUMNS FROM ${safeTable} LIKE ?`, [escapeLike(col.name)]);
          if (cols.length === 0) {
            await query(`ALTER TABLE ${safeTable} ADD COLUMN ${safeCol} ${col.def}`);
            console.log(`Added column ${table}.${col.name}`);
          }
        } catch (err) {
          console.error(`Migration error ${table}.${col.name}:`, err.message);
        }
      }
    } catch (err) {
      console.error(`Migration error for table ${table}:`, err.message);
    }
  }

  for (const c of constraints) {
    try {
      await query(c.sql);
      console.log(`Applied constraint: ${c.name}`);
    } catch (err) {
      if (process.env.NODE_ENV !== 'production') {
        console.log(`Skipped constraint ${c.name}: ${err.message}`);
      }
    }
  }
}

File: internal/database/migrate.go
package database

import (
	"fmt"
	"log"
	"strings"
)

type columnDef struct {
	Name string
	Def  string
}

var tables = map[string][]columnDef{
	"users": {
		{Name: "id", Def: "INT AUTO_INCREMENT PRIMARY KEY"},
		{Name: "email", Def: "VARCHAR(255) NOT NULL"},
		{Name: "username", Def: "VARCHAR(255) NOT NULL"},
		{Name: "password_hash", Def: "VARCHAR(255) NOT NULL"},
		{Name: "ptero_user_id", Def: "INT"},
		{Name: "ptero_uuid", Def: "VARCHAR(255)"},
		{Name: "first_name", Def: "VARCHAR(255)"},
		{Name: "last_name", Def: "VARCHAR(255)"},
		{Name: "password_set", Def: "TINYINT(1) NOT NULL DEFAULT 0"},
		{Name: "is_admin", Def: "TINYINT(1) NOT NULL DEFAULT 0"},
		{Name: "restricted", Def: "TINYINT(1) NOT NULL DEFAULT 0"},
		{Name: "auth_restricted", Def: "TINYINT(1) NOT NULL DEFAULT 0"},
		{Name: "token_version", Def: "INT NOT NULL DEFAULT 0"},
		{Name: "avatar", Def: "VARCHAR(255) DEFAULT NULL"},
		{Name: "ptero_client_api_key", Def: "VARCHAR(255) DEFAULT NULL"},
		{Name: "email_verified", Def: "TINYINT(1) NOT NULL DEFAULT 0"},
		{Name: "verification_token", Def: "VARCHAR(64) DEFAULT NULL"},
		{Name: "verification_token_expires", Def: "DATETIME DEFAULT NULL"},
		{Name: "pending_email", Def: "VARCHAR(255) DEFAULT NULL"},
		{Name: "email_change_token", Def: "VARCHAR(64) DEFAULT NULL"},
		{Name: "email_change_code", Def: "VARCHAR(10) DEFAULT NULL"},
		{Name: "email_change_expires", Def: "DATETIME DEFAULT NULL"},
		{Name: "user_agent", Def: "VARCHAR(512) DEFAULT NULL"},
		{Name: "onboarding_done", Def: "TINYINT(1) NOT NULL DEFAULT 0"},
		{Name: "totp_secret", Def: "VARCHAR(255) DEFAULT NULL"},
		{Name: "totp_enabled", Def: "TINYINT(1) NOT NULL DEFAULT 0"},
		{Name: "recovery_codes", Def: "TEXT DEFAULT NULL"},
		{Name: "created_at", Def: "TIMESTAMP DEFAULT CURRENT_TIMESTAMP"},
	},
	"server_meta": {
		{Name: "id", Def: "INT AUTO_INCREMENT PRIMARY KEY"},
		{Name: "ptero_server_id", Def: "INT NOT NULL"},
		{Name: "user_id", Def: "INT NOT NULL"},
		{Name: "created_at", Def: "TIMESTAMP DEFAULT CURRENT_TIMESTAMP"},
		{Name: "expires_at", Def: "TIMESTAMP NOT NULL"},
		{Name: "status", Def: "ENUM('active','suspended','expired') DEFAULT 'active'"},
		{Name: "suspend_reason", Def: "TEXT DEFAULT NULL"},
		{Name: "suspended_by", Def: "VARCHAR(20) DEFAULT NULL"},
	},
	"user_ips": {
		{Name: "id", Def: "INT AUTO_INCREMENT PRIMARY KEY"},
		{Name: "user_id", Def: "INT NOT NULL"},
		{Name: "ip_address", Def: "VARCHAR(45) NOT NULL"},
		{Name: "user_agent", Def: "VARCHAR(512) DEFAULT NULL"},
		{Name: "created_at", Def: "TIMESTAMP DEFAULT CURRENT_TIMESTAMP"},
	},
	"activity_log": {
		{Name: "id", Def: "INT AUTO_INCREMENT PRIMARY KEY"},
		{Name: "user_id", Def: "INT NOT NULL"},
		{Name: "action", Def: "VARCHAR(50) NOT NULL"},
		{Name: "details", Def: "VARCHAR(255) DEFAULT ''"},
		{Name: "server_id", Def: "INT DEFAULT NULL"},
		{Name: "created_at", Def: "TIMESTAMP DEFAULT CURRENT_TIMESTAMP"},
	},
	"nests": {
		{Name: "id", Def: "INT AUTO_INCREMENT PRIMARY KEY"},
		{Name: "ptero_nest_id", Def: "INT NOT NULL UNIQUE"},
		{Name: "name", Def: "VARCHAR(255) NOT NULL"},
		{Name: "logo", Def: "VARCHAR(255) DEFAULT NULL"},
		{Name: "description", Def: "TEXT DEFAULT NULL"},
		{Name: "unavailable", Def: "TINYINT(1) NOT NULL DEFAULT 0"},
		{Name: "created_at", Def: "TIMESTAMP DEFAULT CURRENT_TIMESTAMP"},
	},
	"egg_resources": {
		{Name: "id", Def: "INT AUTO_INCREMENT PRIMARY KEY"},
		{Name: "ptero_nest_id", Def: "INT NOT NULL"},
		{Name: "ptero_egg_id", Def: "INT NOT NULL"},
		{Name: "logo", Def: "VARCHAR(255) DEFAULT NULL"},
		{Name: "cpu_limit", Def: "INT DEFAULT NULL"},
		{Name: "memory_limit", Def: "INT DEFAULT NULL"},
		{Name: "disk_limit", Def: "INT DEFAULT NULL"},
		{Name: "unavailable", Def: "TINYINT(1) NOT NULL DEFAULT 0"},
		{Name: "created_at", Def: "TIMESTAMP DEFAULT CURRENT_TIMESTAMP"},
		{Name: "updated_at", Def: "TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP"},
	},
	"notifications": {
		{Name: "id", Def: "INT AUTO_INCREMENT PRIMARY KEY"},
		{Name: "user_id", Def: "INT NOT NULL"},
		{Name: "title", Def: "VARCHAR(255) NOT NULL"},
		{Name: "message", Def: "TEXT NOT NULL"},
		{Name: "type", Def: "VARCHAR(20) NOT NULL DEFAULT 'info'"},
		{Name: "link", Def: "VARCHAR(255) DEFAULT NULL"},
		{Name: "is_read", Def: "TINYINT(1) NOT NULL DEFAULT 0"},
		{Name: "created_at", Def: "TIMESTAMP DEFAULT CURRENT_TIMESTAMP"},
	},
	"passkeys": {
		{Name: "id", Def: "INT AUTO_INCREMENT PRIMARY KEY"},
		{Name: "user_id", Def: "INT NOT NULL"},
		{Name: "credential_id", Def: "VARCHAR(512) NOT NULL"},
		{Name: "public_key", Def: "TEXT NOT NULL"},
		{Name: "counter", Def: "INT NOT NULL DEFAULT 0"},
		{Name: "transports", Def: "VARCHAR(255) DEFAULT NULL"},
		{Name: "name", Def: "VARCHAR(255) DEFAULT NULL"},
		{Name: "last_used_at", Def: "TIMESTAMP NULL DEFAULT NULL"},
		{Name: "user_handle", Def: "VARCHAR(255) DEFAULT NULL"},
		{Name: "created_at", Def: "TIMESTAMP DEFAULT CURRENT_TIMESTAMP"},
	},
	"node_settings": {
		{Name: "id", Def: "INT AUTO_INCREMENT PRIMARY KEY"},
		{Name: "ptero_node_id", Def: "INT NOT NULL UNIQUE"},
		{Name: "unavailable", Def: "TINYINT(1) NOT NULL DEFAULT 0"},
		{Name: "created_at", Def: "TIMESTAMP DEFAULT CURRENT_TIMESTAMP"},
		{Name: "updated_at", Def: "TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP"},
	},
}

var constraints = []string{
	"ALTER TABLE server_meta ADD INDEX idx_expires (expires_at)",
	"ALTER TABLE server_meta ADD INDEX idx_user (user_id)",
	"ALTER TABLE server_meta ADD INDEX idx_status (status)",
	"ALTER TABLE server_meta ADD INDEX idx_ptero_server (ptero_server_id)",
	"ALTER TABLE server_meta ADD CONSTRAINT fk_server_meta_user FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE",
	"ALTER TABLE user_ips ADD INDEX idx_ip (ip_address)",
	"ALTER TABLE user_ips ADD INDEX idx_user (user_id)",
	"ALTER TABLE user_ips ADD CONSTRAINT fk_user_ips_user FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE",
	"ALTER TABLE activity_log ADD INDEX idx_activity_user (user_id)",
	"ALTER TABLE activity_log ADD INDEX idx_activity_created (created_at)",
	"ALTER TABLE activity_log ADD INDEX idx_action (action)",
	"ALTER TABLE activity_log ADD CONSTRAINT fk_activity_user FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE",
	"ALTER TABLE egg_resources ADD UNIQUE INDEX idx_egg_resources_nest_egg (ptero_nest_id, ptero_egg_id)",
	"ALTER TABLE notifications ADD INDEX idx_notif_user (user_id)",
	"ALTER TABLE notifications ADD INDEX idx_notif_user_read (user_id, is_read)",
	"ALTER TABLE notifications ADD INDEX idx_notif_created (created_at)",
	"ALTER TABLE notifications ADD CONSTRAINT fk_notif_user FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE",
	"ALTER TABLE users ADD INDEX idx_email (email)",
	"ALTER TABLE users ADD INDEX idx_username (username)",
	"ALTER TABLE passkeys ADD INDEX idx_passkey_user (user_id)",
	"ALTER TABLE passkeys ADD INDEX idx_passkey_credential (credential_id(255))",
	"ALTER TABLE passkeys ADD CONSTRAINT fk_passkey_user FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE",
	"ALTER TABLE node_settings ADD INDEX idx_node_settings_node (ptero_node_id)",
}

func escapeID(id string) string {
	return "`" + strings.ReplaceAll(id, "`", "``") + "`"
}

func RunMigrations() {
	for table, cols := range tables {
		safeTable := escapeID(table)
		var exists int
		err := DB.QueryRow("SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = DATABASE() AND table_name = ?", table).Scan(&exists)
		if err != nil {
			log.Printf("Migration check error for %s: %v", table, err)
			continue
		}
		if exists == 0 {
			var colDefs []string
			for _, c := range cols {
				colDefs = append(colDefs, escapeID(c.Name)+" "+c.Def)
			}
			_, err := DB.Exec(fmt.Sprintf("CREATE TABLE %s (%s)", safeTable, strings.Join(colDefs, ", ")))
			if err != nil {
				log.Printf("Failed to create table %s: %v", table, err)
			} else {
				log.Printf("Created table: %s", table)
			}
			continue
		}
		for _, col := range cols {
			var colExists int
			err := DB.QueryRow("SELECT COUNT(*) FROM information_schema.columns WHERE table_schema = DATABASE() AND table_name = ? AND column_name = ?", table, col.Name).Scan(&colExists)
			if err != nil {
				continue
			}
			if colExists == 0 {
				_, err := DB.Exec(fmt.Sprintf("ALTER TABLE %s ADD COLUMN %s %s", safeTable, escapeID(col.Name), col.Def))
				if err != nil {
					log.Printf("Migration error %s.%s: %v", table, col.Name, err)
				} else {
					log.Printf("Added column %s.%s", table, col.Name)
				}
			}
		}
	}

	for _, sql := range constraints {
		if _, err := DB.Exec(sql); err != nil {
			log.Printf("Skipped constraint: %v", err)
		}
	}
}
